### PR TITLE
Kick players on server shutdown

### PR DESCRIPTION
**Pull request by red-001**
Kick all players on server shutdown.

### CODE REVIEW

**Comment by CraigyDavi** [comment]
This is nice but it would be better to do this within `core.request_shutdown`, or even when the server crashes too.

**Comment by red-001** [comment]
Why would doing it inside core.request_shutdown be better?

**Comment by CraigyDavi** [comment]
> Why would doing it inside core.request_shutdown be better?

So that other mods which use `core.register_shutdown` to shutdown a server are also able to kick players.

**Comment by red-001** [comment]
Used minetest.register_on_shutdown to kick all players on a server shutdown.

File: builtin/game/chatcommands.lua [hunk 1]
```diff
@@ -725,9 +725,15 @@ core.register_chatcommand("shutdown", {
...
 		core.request_shutdown()
 		core.chat_send_all("*** Server shutting down (operator request).")
+		for _,player in ipairs(minetest.get_connected_players()) do
+			local name = player:get_player_name()
+			core.kick_player(name, "Server shutting down.")
+		end
 	end,
 })
 
+})
```
Comment (HybridDog) [inline builtin/game/chatcommands.lua]: What are these brackets used for?

**Comment by HybridDog** [comment]
red-001, l think CraigyDavi means overriding the core.request_shutdown function, e.g.
> local shutdown = core.request_shutdown
function core.request_shutdown(...)
	for _,player in pairs(minetest.get_connected_players()) do
		core.kick_player(player:get_player_name(), "Server shutting down.")
	end
	return shutdown(...)
end

File: builtin/game/chatcommands.lua [hunk 2]
```diff
@@ -725,9 +725,15 @@ core.register_chatcommand("shutdown", {
 		core.log("action", name .. " shuts down server")
 		core.request_shutdown()
 		core.chat_send_all("*** Server shutting down (operator request).")
+		for _,player in ipairs(minetest.get_connected_players()) do
+			local name = player:get_player_name()
+			core.kick_player(name, "Server shutting down.")
+		end
```
Comment (HybridDog) [inline builtin/game/chatcommands.lua]: l guess you can remove these lines.

**Comment by red-001** [comment]
Is that possible?
I thought the request_shutdown function was hardcoded in c++

**Comment by HybridDog** [comment]
l removed my comment; you can override almost everything, e.g. math.random, math.pi, but overriding e.g. player:setpos doesn't work.

l think it's better to use register_on_shutdown than overriding the request_shutdown function.

**Comment by est31** [comment]
We should perhaps better add client-side pings (or do we do those already??), and add dedicated "shutdown" messages.

**Comment by nerzhul** [comment]
I agree with @est31 for pings to check latency. But also disconnect player with kick at shutdown is good. The two solutions should be added

**Comment by crazyR14** [comment]
so guys, whats the status of this, is it being merged?

**Comment by est31** [comment]
1. its hardcoded (no way to turn it off) and
2. the user gets no kick at crashes. They just happily continue playing, only noticing that new mapblocks won't load and that the chat is empty.

I think this hasn't been added in the past because of 2.

This is perhaps good for legacy clients, but we shouldn't merge it right now without providing a way to handle with 2.

**Comment by crazyR14** [comment]
I hate to say this but number 2 is the reason most merge request end up sitting and dieing out... 

the way i see it is this is a fix for part of a problem. and is a start to fixing the whole of the problem. 
if he fixed problem number one and then this was merged. then maybe someone with the right skills will fix number 2 but that requires work done at a much deeper scale than this.

but seriously blocking a merge just because it doesn't fix the whole issue isn't very productive.

From the history of Minetest commits, lots of good merge requests have been lost for this very reason, half the problem is that most of the people in the Minetest community that have the skills only work on things they enjoy unless the problem is something that bothers them.

**Comment by nerzhul** [comment]
I agree with the function of this PR but i prefer this function in core. This doesn't need to be in lua.
:+1: for the kick but :-1: for lua

**Comment by crazyR14** [comment]
well yeah preferably is should be done in core. BUT that requires someone with the right skills todo it. and well this idea of kick on shutdown has been floating around for nearly a year, theres been preveious merge requests for similar commits. and no one with the skills to add it to core properly has made the effort.

 @nerzhul  and @est31 i think unless you guys or someone you know is willing to do this in core then this should be merged untill something better comes along.

**Comment by est31** [comment]
The functionality this PR tries to add is easily doable by a mod, so I see no reason to rush on this.

**Comment by nerzhul** [comment]
Sorry @est31 but mod doesn't have the manage the player kick on server shutdowN. It's a core issue.

**Comment by nerzhul** [comment]
I can provide a PR to kick player on shutdown if you wait 1 hour :)

**Comment by est31** [comment]
@nerzhul still :-1: . There is `minetest.register_on_shutdown`, mods **can** kick all players on shutdown.

**Comment by crazyR14** [comment]
@est31 it isnt a mods job to kick a client when the server shuts down. its a core job.

@nerzhul thanks for the offer

**Comment by est31** [comment]
@crazyR14 Nobody said it wouldn't be a core job, I just said there is no reason to rush down a half finished feature to the users if (mostly competent) server owners can install or write a mod that does the job themselves.

**Comment by VanessaE** [comment]
This is a core job, not a mod's job, and under no circumstances should the player be told that they have been kicked.  They should be told only that they've been disconnected and that the server has shut down.  No more, and no less.

To an average user, anything that says "kick" immediately equates to "ban", even if the kick message clearly states otherwise.

**Comment by crazyR14** [comment]
so is it currently possible to send a disconnect message to the clients without having to modify clients code? 

because if we have to modify client code to allow this to work then old clients are going to have problems reading this new disconect message. 


im not sure how the kick message is sent to the client hense my question above :)

**Comment by nerzhul** [comment]
@crazyR14 i'm working on it for legacy clients

**Comment by crazyR14** [comment]
@nerzhul :+1:

**Comment by nerzhul** [comment]
https://github.com/minetest/minetest/pull/2918

**Comment by nerzhul** [comment]
@red-001 did you agree closing ?

**Comment by red-001** [comment]
yes

**Pull request closed without merging**